### ISSUE TITLE
Avoid updateNeighborListSize on smaller systems

### ISSUE THREAD

**Issue by bdenhollander** [openmm/openmm#4272]
Queuing and waiting for events results in idle GPU time for AMD on OpenCL on Windows. For smaller systems (~<50000 atoms), would it be a viable optimization to allocate the maximum amount of memory needed to store `interactingTiles` and `interactingAtoms`? Then setting up `downloadCountEvent` and checking could be skipped since `updateNeighborListSize` would no longer be needed. The arrays would already be at their maximum size and be able to hold all the interactions.

I ran a few quick tests and calculated the memory in MB needed as:

```C++
int maxTiles = numBlocks * (numBlocks + 1) / 2;
int memoryMB = (maxTiles + OpenCLContext::TileSize * maxTiles) * sizeof(cl_int) / 1024 / 1024
```

Results for AMD RX 6600 on Windows 10 OpenCL 3584.0

Water boxes with 1.0 cutoff

| Width | Atoms | Memory (MB) | Original (s) | New (s) | Ratio |
|-------|-------|-------------|--------------|---------|-------|
| 4     | 6282  | 2           | 2.96         | 1.53    | 0.52  |
| 5     | 12255 | 9           | 4.07         | 2.50    | 0.62  |
| 6     | 21384 | 28          | 4.75         | 3.09    | 0.65  |
| 7     | 33840 | 70          | 6.33         | 4.50    | 0.71  |
| 8     | 50442 | 156         | 8.56         | 6.67    | 0.78  |
| 9     | 72279 | 321         | 11.41        | 9.58    | 0.84  |

As memory usage grows, the speed improvement decreases since more atoms means more time in computation rather than communication.

**Comment by bdenhollander** [comment]
I've cleaned up my test code and added a memory size check instead of using an arbitrary number of atoms. Faster cards tend to also have more VRAM so this should scale.
https://github.com/bdenhollander/openmm/tree/skip-resize-neighbor-list

| Test              | Memory (MB) | Original (ns/day) | New (ns/day) | Ratio |
|-------------------|-------------|-------------------|--------------|-------|
| gbsa              | 0.4         | 466.521           | 654.195      | 1.40  |
| rf                | 34.2        | 481.068           | 661.339      | 1.37  |
| pme               | 34.2        | 312.645           | 435.447      | 1.39  |
| apoa1rf           | 523.0       | 140.562           | 165.339      | 1.18  |
| apoa1pme          | 523.0       | 94.2161           | 107.613      | 1.14  |
| apoa1ljpme        | 523.0       | 72.7853           | 81.3295      | 1.12  |
| amoebagk          |              | 1.07399           | 1.07372      | 1.00  |
| amoebapme         | 68.5        | 1.09125           | 1.08664      | 1.00  |
| amber20-dhfr      | 34.2        | 333.968           | 472.123      | 1.41  |
| amber20-cellulose |             | 23.9454           | 23.9407      | 1.00  |
| amber20-stmv      | 363.6       | 7.09299           | 7.19104      | 1.01  |

**Comment by peastman** [comment]
That's an interesting idea.  Let me run some tests on other GPUs and see if I find a similar difference.  We also should try to figure out why it's having such a big impact, and whether there's anything else that can be done to reduce it.

**Comment by peastman** [comment]
Here's what I get on M1 Pro (OpenCL) and RTX 4080 (OpenCL and CUDA).

|GPU|Platform|Test|Original|New|Ratio|
|---|---|---|---|---|---|
M1 Pro|OpenCL|pme|157|160|1.02
M1 Pro|OpenCL|apoa1pme|38|39|1.00
RTX 4080|OpenCL|pme|1473|1411|0.96
RTX 4080|OpenCL|apoa1pme|477|483|1.01
RTX 4080|CUDA|pme|2047|2124|1.04
RTX 4080|CUDA|apoa1pme|631|627|0.99

It has very little effect in all cases.  Maybe there's a problem that's specific to AMD?

**Comment by bdenhollander** [comment]
The optimization works by eliminating idle GPU gap introduced by `downloadCountEvent.wait()` after `computeNonbonded` as shown here in Radeon GPU Profiler.

![image](https://github.com/openmm/openmm/assets/44237618/fe6745a1-4f4a-44a1-aba8-61080a5ada6b)

Those gaps were mostly mitigated on Apple by #3960. Whether those gaps exist on NVIDIA + Windows or Intel is unknown. Anecdotally, some low atom count FAH projects have at times been restricted to only run on Linux + NVIDIA GPUs since performance was poor on Windows. I would expect to see an improvement on your [RX 5700 XT](https://github.com/openmm/openmm/issues/3937#issuecomment-1412702988) in Linux.

Eliminating the gaps is similar to adding [flush](https://github.com/openmm/openmm/pull/3973#issuecomment-1703822145) to reduce them for OpenCL on Windows. Performance is slightly better since the GPU-host transfer is eliminated completely.

**Comment by peastman** [comment]
There shouldn't be a gap.  The transfer is done asynchronously.  Your profile shows that over 500 us pass between when it's started and when `computeNonbonded()` completes.  That should be more than enough time to transfer just a few bytes from the device back to the host.  The transfer should complete and `wait()` should return long before the GPU runs out of work.  So what is creating a gap of several hundred us?

**Comment by peastman** [comment]
I tried RX 5700 XT on Linux.

|Test|Old|New|Ratio|
|---|---|---|---|
pme|330|407|1.23
apoa1pme|103|114|1.10


There's a large speedup for pme and a smaller speedup for apoa1pme.

With Apple GPUs we already call `flush()` immediately after starting the copy and immediately before calling `wait()`.  I tried doing the same on AMD, but the effect was negligible.

**Comment by bdenhollander** [comment]
AMD on Linux likes one `flush()` just before the call to `computeNonbonded`.
https://github.com/openmm/openmm/issues/3937#issuecomment-1413872621
https://github.com/openmm/openmm/issues/3937#issuecomment-1416377734

**Comment by peastman** [comment]
Adding a flush just before computeNonbonded also helps pme a lot (399 ns/day), but has no benefit for apoa1pme.

I also tried adding only the second of the two flushes used on Mac.  That helps both benchmarks, though not as much as completely removing the check: 162 for pme and 106 for apoa1pme.

**Comment by bdenhollander** [comment]
I modified #3973 to apply the flush optimization only to AMD on Windows and Linux since there has been no feedback from Nvidia testers on Windows. It would be great to get that into an OpenMM 8.1 RC.

**Comment by peastman** [comment]
On AMD should the flush go just before `computeNonbonded()` rather than before `wait()`?  In my test that gave a bigger speedup for pme.  It had no benefit for apoa1pme, but the speedup from the other one is pretty minimal in that case too, just a few percent.

**Comment by bdenhollander** [comment]
I was looking at the 104 for apoa1pme in https://github.com/openmm/openmm/issues/3937#issuecomment-1416377734 for perm 5 without realizing that it wasn't one of the best results for pme. The flush has been moved to before `computeNonbonded`.

**Comment by peastman** [comment]
I'm closing this, since the issue is addressed by #3973.
